### PR TITLE
chore: Cleanup setup and base css

### PR DESCRIPTION
**Pull request by eslachance**
## Context
This PR makes some cleanup in the theme and layout, to start updating the site code to more modern standards. 

There's still a lot of problems and downsides to using Hugo, but at least, now we're less based on 1997 standards. Or at least, it's a start. 

## Preview
Can be viewed on [cloudflare](https://foulab-dot-org-hugo.pages.dev/).

## Screenshot:
image

## Things that were changed: 

- Flex is used for the main layout, rather than floats
- <center> is long gone (from the layout itself, can't remove it from the markdown content)
- Consistent indentation within the source files (the html output isn't significant)
- Move to html5 standards. this isn't 1997, XHTML 1.0 Transitional isn't a thing anymore. (removed associated "badge")
- Added Mattermost link (let me know if this shouldn't be public)
- Very slight stylistic adjustments in the nav, footer, etc. Nothing major.

### CODE REVIEW

**Comment by cpatulea** [comment]
Awesome! Thanks a lot for working on this. I will need some time (a few days) to look at this. In the mean time, I also added some people who are closely involved with the site and might want to review.

File: themes/foulab-2018/layouts/_default/baseof.html [hunk 1]
```diff
@@ -1,73 +1,40 @@
...
+<html lang="en">
   <head>
+	  <meta charset="utf-8">
 	  <title>{{ block "title" . }}{{ default .GitInfo.Subject .Title | default "Welcome to Foulab! - foulab.org" }}{{ end }}</title>
     
-    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
-  
-    <link rel="stylesheet" href="{{ .Site.BaseURL }}/css/default/default.css" />
-    <link rel="stylesheet" href="{{ .Site.BaseURL }}/css/default/app.css" />
+    <link rel="stylesheet" href="/css/default/default.css">
```
Comment (cpatulea) [inline themes/foulab-2018/layouts/_default/baseof.html:7]: Hmm, I'm not a Hugo expert, so I did a bit of research what's the idiomatic way to reference assets.

`{{ .Site.BaseURL }}/foo` is definitely *not* the idiomatic way, so thanks for fixing it.

I think the idiomatic way is resources.Get + r.RelPermalink, example: https://github.com/luizdepra/hugo-coder/blob/52b14a50f716cb7e69df8d0804f1b68aeec8a08c/layouts/_default/baseof.html#L41

In the short term, it doesn't give us any immediate value. But long-term:
- For JS assets, it makes it very easy to enable minification, if we want.
- For image assets, resizing
- For all asset types, resource integrity

I think it would also make it easier to serve drafts of the site on non-root URLs (eg example.com/foulab-draft/) just by changing a path in a single config file.

what do you think about switching to resources API throughout?

Maybe let's wait for @0x5c to also chime in.

File: layouts/index.html [hunk 2]
```diff
@@ -4,27 +4,7 @@
 {{ partial "article.html" . }}
 {{ end }}
 
-<a href="./news/">» See all articles</a>
-
-<script>
```
Comment (cpatulea) [inline layouts/index.html:9]: I had actually put the ical script inline on purpose, to avoid a network round-trip, for case where a human is visiting, which is latency-sensitive and the most common case.

File: themes/foulab-2018/layouts/partials/header.html [hunk 3]
```diff
@@ -1,39 +1,41 @@
...
-	</div>
-	
-      </div>
-      
+  <div class="tinybar">
+
+    <a href="https://github.com/FOULAB" title="Foulab GitHub">
+      <img class="icons" src="/img/git.png" alt="Link to Foulab GitHub">
+    </a>
+    <a href="https://test.foulab.org/" title="Mattermost Server">
```
Comment (cpatulea) [inline themes/foulab-2018/layouts/partials/header.html]: I'm on the fence about linking to Mattermost.
- for the majority of visitors to the site (non-members, people who have never been to the space), they won't be able to log in, so it won't be useful and would probably just raise questions
- Mattermost is a bit of an 'open secret' for people who have visited the lab in person, and I think that's a nice thing.

@kryma-foulab @0x5c what do you think?

File: themes/foulab-2018/layouts/shortcodes/centered-image.html [hunk 4]
```diff
@@ -1,3 +1,7 @@
-<CENTER>
-  <img src="{{ .Site.BaseURL }}/media/image/{{ .Page.File.Dir }}{{ .Page.File.BaseFileName }}/{{ .Get "src" }}" {{ with .Get "height" }}height="{{ . }}"{{ end }} {{ with .Get "width" }}width="{{ . }}"{{ end }} />
-</CENTER>
+<img
+  src="/media/image/{{ replace .Page.File.Dir "\\" "/"}}{{ .Page.File.BaseFileName }}/{{ .Get "src" }}"
```
Comment (cpatulea) [inline themes/foulab-2018/layouts/shortcodes/centered-image.html:2]: What's the reason for this 'replace'? Is it because you were developing on Windows and it was necessary to construct a valid path?

**Review by cpatulea — CHANGES_REQUESTED** [review]
Overall, looks great, thanks a lot for doing this.

I checked it on desktop and mobile (iPhone Chrome, small screen) and it looks good.

File: themes/foulab-2018/layouts/shortcodes/centered-image.html (hunk 4 above)
Comment (eslachance) [reply themes/foulab-2018/layouts/shortcodes/centered-image.html:2]: Actually yes, there was some weirdness with the Dir having a backslash inside of it - so I was getting a path that was `%5c` instead of forward slashes.

File: themes/foulab-2018/layouts/partials/header.html (hunk 3 above)
Comment (eslachance) [reply themes/foulab-2018/layouts/partials/header.html]: Yeah I wasn't sure about this one either - I can easily revert this if the consensus is that it shouldn't be there. Personally I had no idea what the URL was and needed to login from my phone at one point, so this would have helped me... but it's probably a rare occurrence.

File: layouts/index.html (hunk 2 above)
Comment (eslachance) [reply layouts/index.html:9]: Understood - will revert.

File: themes/foulab-2018/layouts/partials/header.html (hunk 3 above)
Comment (kryma-foulab) [reply themes/foulab-2018/layouts/partials/header.html]: Probably not a good idea to put Mattermost - although it is a visual treat to see the thing in the sidebar :)
If you'd like to get fancy with it, is there any good way to serve it only to the lab's IP?

File: themes/foulab-2018/layouts/_default/baseof.html (hunk 1 above)
Comment (kryma-foulab) [reply themes/foulab-2018/layouts/_default/baseof.html:7]: Would it work with a local deployment? A potential developer's workflow could be putting `file:///home/user/foulab-dot-org-hugo/public/index.html` into the address bar, as of right now it works pretty well with a slight config override.

File: themes/foulab-2018/layouts/partials/header.html [hunk 5]
```diff
@@ -1,39 +1,41 @@
...
+    </a>
+    <a href="https://twitter.com/foulab" title="Twitter">
+      <img class="icons" src="/img/twt.png" alt="Link to Twitter">
+    </a>
+    <a href="/news/index.xml" title="RSS Feed">
+      <img class="icons" src="/img/rss.png" alt="Link to RSS Feed">
+    </a>
+    <a href="mailto:friends.of.foulab-subscribe@lists.riseup.net" title="Mailing List">
+      <img class="icons" src="/img/eml.png" alt="Link to Mailing List">
+    </a>
```
Comment (kryma-foulab) [inline themes/foulab-2018/layouts/partials/header.html:31]: Is this still active?

File: themes/foulab-2018/static/js/status.js [hunk 6]
```diff
@@ -1,12 +1,9 @@
...
-
-document.addEventListener("DOMContentLoaded", function(event) {
-    var a = new XMLHttpRequest();
-    a.addEventListener("load", handleStatus);
-    a.open("GET", "https://foulab.org/status/");
-    a.send();
-});
+// Fetch the status of the lab from the server and display it.
+// Thank you modern javascript for making this so easy!
+fetch("https://foulab.org/status/")
```
Comment (kryma-foulab) [inline themes/foulab-2018/static/js/status.js]: Should this be relative?

File: themes/foulab-2018/layouts/partials/header.html (hunk 3 above)
Comment (eslachance) [reply themes/foulab-2018/layouts/partials/header.html]: Hiding it visually might be possible but this would be client-side so it would still be viewable by someone looking at the source code, so not really a good solution.

File: themes/foulab-2018/layouts/partials/header.html (hunk 5 above)
Comment (eslachance) [reply themes/foulab-2018/layouts/partials/header.html:31]: Definitely not a question I can personally answer 😓

File: themes/foulab-2018/static/js/status.js (hunk 6 above)
Comment (eslachance) [reply themes/foulab-2018/static/js/status.js]: Definitely can be, yes. It doesn't work locally due to CORS anyways, so relative or not doesn't make much difference. I'll change it.

File: themes/foulab-2018/layouts/partials/header.html (hunk 3 above)
Comment (cpatulea) [reply themes/foulab-2018/layouts/partials/header.html]: Hiding Mattermost does not need to be 'secure', in my opinion, (heck it might even be a fun game that you get access if you view the source code and solve a puzzle ;-) ) but I just think it would be brittle and tedious to maintain an IP allowlist. So I suggest simply leave it out.

File: themes/foulab-2018/layouts/shortcodes/centered-image.html (hunk 4 above)
Comment (cpatulea) [reply themes/foulab-2018/layouts/shortcodes/centered-image.html:2]: I'm not sure it's a good idea to have OS-specific code in the templates. I have a hunch that Hugo "resources.Get + r.RelPermalink" (as discussed in the other thread) would solve the backslash problem. Would you mind trying that?

File: themes/foulab-2018/layouts/partials/header.html (hunk 5 above)
Comment (cpatulea) [reply themes/foulab-2018/layouts/partials/header.html:31]: Looks like it's been there since inception of the Hugo version of the site, 6 years ago: https://github.com/FOULAB/foulab-dot-org-hugo/blob/45b7ad806fdce3216f499403b71c69d98183f87f/themes/foulab-2018/layouts/partials/header.html#L19

I don't have any tacit knowledge about this, I don't have admin access to that mailing list, if it even still exists.

So I suggest deleting this. Thanks for checking.

File: themes/foulab-2018/static/js/status.js (hunk 6 above)
Comment (cpatulea) [reply themes/foulab-2018/static/js/status.js]: Looks like this has been there since inception: https://github.com/FOULAB/foulab-dot-org-hugo/blob/45b7ad806fdce3216f499403b71c69d98183f87f/themes/foulab-2018/static/js/status.js#L10

I think absolute might help with local testing, because if the lab happens to the open, it exercises the code path and makes the OPEN element show.

I think you're right CORS might be breaking this right now. I'm happy to weak the CORS headers on /status/ to allow any origin fetch it.

File: themes/foulab-2018/layouts/_default/baseof.html (hunk 1 above)
Comment (eslachance) [reply themes/foulab-2018/layouts/_default/baseof.html:7]: I've tried a few things and I'm basically running into different problems every time. 

The use of a variable as per the github link just doesn't work at all
```
    {{ $script := resources.Get "js/status.js" }}
    <script src="{{ $script.RelPermalink }}"></script>
```
This gives me an error about `nil pointer evaluating resource.Resource.RelPermalink`

I found alternatives for Site.BaseURL that will basically give me the same problem it did. There's something wrong in Hugo itself, because BaseURL as well as urls.absUrl just somehow don't work as expected.

So, for example, with the following code: 
```
    <link rel="stylesheet" href="{{ absLangURL "css/default/default.css" }}">
    <link rel="stylesheet" href="{{ absLangURL "css/default/app.css" }}">
```
And running `hugo serve -p 3000` I get a proper server running on `http://localhost:3000` but the HTML in the head just... don't care. 
image

Running with a `--baseURL=https://c0c01e372d48.ngrok.app/` for example, seems to work except that it doesn't, it forcefully appends the port to the end: 
image

And without a port specified, letting the default, I just get a default port like `https://c0c01e372d48.ngrok.app:62492/` (not sure when it decided to move away from 1313 but it sure did decide this), and I can't really test whether this would work or not since rebinding ngrok to a different port generates a new address so now I'm sort of stuck with an infinite loop of "not working".

All in all, my point here is that Hugo's system is pretty broken, and honestly, the only situation right now where this would matter would be if the baseURL somehow isn't on the root of the domain. BaseURLs really only matter when you're on some subfolder in a more complex configuration or if you're on a share hosting where you can't be on the root.

Basically, getting frustrated over trying to go around Hugo's broken system just to fix a non-problem isn't how I want to spend my days 🤣

Comment (eslachance) [reply themes/foulab-2018/layouts/_default/baseof.html:7]: > Would it work with a local deployment? A potential developer's workflow could be putting `file:///home/user/foulab-dot-org-hugo/public/index.html` into the address bar, as of right now it works pretty well with a slight config override.

Using `file://` to locally develop isn't a thing we do in 2024. There are many, *many* trivial ways to start a proper thin local web server on a local machine, from using python's `SimpleHTTPServer` module, to VSCode's Live Server, to a ton of single-executable light http servers. This is a non-problem.

**Comment by cpatulea** [comment]
@0x5c should we wait for you to look at this and weigh in; or [if we iron out all the kinks] you don't mind if we merge without your feedback?

File: .vscode/settings.json
Comment (0x5c) [inline .vscode/settings.json:1]: Would be best to avoid commiting local editor-specific configs to a public repo.
It will create conflicts for anyone who has to locally edit theirs

For those settings which are relevant to style, using [EditorConfig](https://editorconfig.org) is the normal way to automatically communicate them to editors. It is supported out of the box by numerous editors, including VSCode

**Comment by 0x5c** [comment]
> @0x5c should we wait for you to look at this and weigh in; or [if we iron out all the kinks] you don't mind if we merge without your feedback?

I'm getting to it at the moment

File: themes/foulab-2018/static/js/shopifydonation.js
Comment (0x5c) [inline themes/foulab-2018/static/js/shopifydonation.js:1]: Shouldn't be a blocker for merging, but the question of whether or not (and if so, why) shopify is still needed despite having Zeffy now should be raised with the funding/etc people

File: themes/foulab-2018/static/img/patreon-logo.svg
Comment (0x5c) [inline themes/foulab-2018/static/img/patreon-logo.svg:1]: I recall hearing some time ago that the lab was trying to move away from patreon, something about payment problems and unneeded complexity. Same as the Shopify code, shouldn't be a blocker but should be raised with the funding people.

File: content/about.md
Comment (0x5c) [inline content/about.md:1]: At least locally, this page is broken with this PR; everything after the video itself is gone from rendered output

image

File: themes/foulab-2018/layouts/_default/baseof.html (hunk 1 above)
Comment (0x5c) [reply themes/foulab-2018/layouts/_default/baseof.html:7]: > many trivial ways to start a proper thin local web server on a local machine

Most trivially, considering we already have to use that tool to build the site; `hugo serve` which does both the site regeneration and serving, with live updates too

It looks though like there might be a bug in hugo or holes in our config/shortcodes; https://github.com/gohugoio/hugo/issues/8896 [and related forum thread](https://discourse.gohugo.io/t/resources-get-throw-error-if-argument-is-a-shortcode-parameter/34299)

File: themes/foulab-2018/static/css/default/app.css [hunk 7]
```diff
@@ -1,23 +1,69 @@
...
+  color: #000000;
+  font-size: 10pt;
+  padding: 10px;
+  flex-basis: 0;
+  flex-grow: 999;
+}
+
+#nav {
+  width: 150px;
+  background-color: grey;
```
Comment (0x5c) [inline themes/foulab-2018/static/css/default/app.css:43]: This new background colour makes the nav fail even more on the WCAG contrast guidelines for text
[Current: `4.6:1`](https://webaim.org/resources/contrastchecker/?fcolor=FFFFFF&bcolor=757575)
[The PR: `3.94:1`](https://webaim.org/resources/contrastchecker/)

File: themes/foulab-2018/layouts/partials/header.html (hunk 5 above)
Comment (0x5c) [reply themes/foulab-2018/layouts/partials/header.html:31]: Looking on riseup's list server, it seems there's no longer a list with that name. Should be safe to remove

File: themes/foulab-2018/layouts/partials/navigation.html
Comment (0x5c) [inline themes/foulab-2018/layouts/partials/navigation.html:1]: On pages with content shorter than the nav, the page layout shinks to fit the exact contents of the nav. It gets a bit claustrophobic

old 
image
new 
image

File: themes/foulab-2018/static/css/default/app.css
Comment (0x5c) [inline themes/foulab-2018/static/css/default/app.css:1]: Is there a specific reason that the site has two CSS files, both seemingly applied globally? This setup dates from the start of the hugo-based website, but is there anything obvious that would prevent merging them? (should not be a blocker)

File: themes/foulab-2018/layouts/_default/baseof.html
Comment (0x5c) [inline themes/foulab-2018/layouts/_default/baseof.html:1]: With the PR there's a weird glitch with the page/nav/header width, which also deletes the right border of the nav

old
image


new
image

**Review by 0x5c — CHANGES_REQUESTED** [review]

**Comment by cpatulea** [comment]
Hey, still planning on working on this?
